src/main/scala/firrtl/passes/ExpandWhens.scala

```diff
@@ -133,7 +133,11 @@ object ExpandWhens extends Pass {
                     case (WInvalid(), WInvalid()) => WInvalid()
                     case (WInvalid(), fv) => ValidIf(NOT(s.pred), fv, tpe(fv))
                     case (tv, WInvalid()) => ValidIf(s.pred, tv, tpe(tv))
-                    case (tv, fv) => Mux(s.pred, tv, fv, mux_type_and_widths(tv, fv))
+                    case (tv, fv) => {
+                      // parameterization might result in the same tv, fv -- get rid of unnecessary mux
+                      if (tv != fv) Mux(s.pred, tv, fv, mux_type_and_widths(tv, fv))
+                      else tv
```

Comment:
I agree with @jackkoenig that we should consider rerunning ConstProp, CSE, and DCE. We should also remember that these should never be required for correctness.

> If basic constant propagation shouldn't be done here, then there definitely needs to be some utility mechanism to traverse the graph and figure out if 2 nodes have the same origin.

Perhaps, but you shouldn't care whether the wmask/enable are equivalent, because your code should work regardless. Simple passes are better passes. We should really just take a look at improving ConstProp so all cases of the above are caught, not just for your use case of wmasks/enables.

> I personally dislike the notion of always having to wait until the end to do constant propagation, because that makes intermediate FIRRTL code harder to decipher and potentially harder to do anything with. (If we want some conditional optimization).

If your pass is too hard to look at, then you should run the ConstProp after your pass when debugging. Don't make ConstProp a necessary thing to run because other users may not want to have any ConstProp done on their code, but still want to use your pass (like @jackkoenig  said, separation of concerns)

> Verilog readibility + code line reduction is obviously still quite valuable to the hardware designer...

Absolutely! And if it isn't readable, we should do a better job.

> so wherever possible, my opinion is you should try to sneak in optimizations that make sense.

I totally disagree. Imagine trying to debug some code where some random pass is doing constant propagation, and I have no way of knowing which one. Sneaking in optimizations is an anti-pattern. I want to be able to turn on/off ALL ConstProp etc. easily, and if these are snuck in everywhere, that is impossible.

W.r.t. this change, I can see this specific optimization in ExpandWhens to _maybe_ be ok because, in a different light, its essentially redefining the behavior of ExpandWhens to be 'insert a mux if the values are different, otherwise return the value'. However, I agree with Jack that we definitely want a separation of concerns, and in general we should NOT add constprop/other optimizations in compiler passes that aren't specifically dedicated to that purpose. My vote is it leaves a sour taste in my mouth but I'm ok merging it. However, I'll leave this decision up to @jackkoenig ;)